### PR TITLE
8301819: Enable continuations code by default

### PR DESCRIPTION
**Pull request by AlanBateman**
The continuations code added in JDK 19 with JEP 425 is currently disabled by default and is only enabled when running with preview features enabled (`--enable-preview`). Disabled by default was the pragmatic choice at that time to reduce risk and possible performance regressions. We'd like to change this so the code is enabled by default on the platforms that have a port. Enabling this code by default helps to shake out any issues early in JDK 21 and sets things up to make virtual threads a permanent feature in the future. If something comes up in the next few weeks/months then it can be changed to be disabled by default again. This change has no impact on the builds that do not have a port.

Testing: tier1-tier6

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301819](https://bugs.openjdk.org/browse/JDK-8301819): Enable continuations code by default

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [2ef27802](https://git.openjdk.org/jdk/pull/12459/files/2ef27802b22281c499133560b3e1c97ab22366f3)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [2ef27802](https://git.openjdk.org/jdk/pull/12459/files/2ef27802b22281c499133560b3e1c97ab22366f3)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12459/head:pull/12459` \
`$ git checkout pull/12459`

Update a local copy of the PR: \
`$ git checkout pull/12459` \
`$ git pull https://git.openjdk.org/jdk pull/12459/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12459`

View PR using the GUI difftool: \
`$ git pr show -t 12459`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12459.diff">https://git.openjdk.org/jdk/pull/12459.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Alan showed me performance data with nothing alarming (at least from benchmarks we use).
I agree with these changes.

**Comment by dcubed-ojdk** [comment]
@AlanBateman - Thanks for documenting that Tier[1-6] testing was done on this change.

**Review by dholmes-ora — APPROVED** [review]
@AlanBateman I agree with enabling this and giving it time to bake. Vladimir has already checked there are no red flags on the performance front, so that is fine. If anything does arise we have, as you note, time to address it.

Copyright year needs updating in continuation.cpp.

Thanks.

**Comment by AlanBateman** [comment]
> @AlanBateman I agree with enabling this and giving it time to bake. Vladimir has already checked there are no red flags on the performance front, so that is fine. If anything does arise we have, as you note, time to address it.

@kuksenko did a lot of detailed analysis on the performance, this lead to JDK-8300002 to account for the interaction between inlining and the post-call nops. We can re-visit again in a few weeks or months if something comes up.

> Copyright year needs updating in continuation.cpp.

Yes, will do.

File: src/hotspot/share/runtime/continuation.cpp [hunk 1]
```diff
@@ -411,9 +411,8 @@ void Continuations::init() {
   Continuation::init();
 }
 
-// While virtual threads are in Preview, there are some VM mechanisms we disable if continuations aren't used
 bool Continuations::enabled() {
-  return VMContinuations && Arguments::enable_preview();
```
Comment (shqking) [inline src/hotspot/share/runtime/continuation.cpp:416]: With this change, I think header file `"runtime/arguments.hpp"` at line 29 can be excluded.

Comment (AlanBateman) [reply src/hotspot/share/runtime/continuation.cpp:416]: > With this change, I think header file `"runtime/arguments.hpp"` at line 29 can be excluded.

Indeed, that header file is no longer needs to be included - thanks!

File: src/hotspot/share/runtime/continuation.hpp [hunk 2]
```diff
@@ -39,7 +39,7 @@ class RegisterMap;
 class Continuations : public AllStatic {
 public:
   static void init();
-  static bool enabled(); // TODO: used while virtual threads are in Preview; remove when GA
+  static bool enabled();
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/continuation.hpp:42]: Based on the comment, I was expecting this line to be removed.
I'll take it that there's still a way to disable Continuations even after virtual
threads come out of "Preview"?

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/continuation.hpp:42]: And now I'm mulling why `VMContinuations` doesn't a default value...

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/continuation.hpp:42]: Ohhh... that's right. It's not implemented on all platforms yet... got it...

**Review by dcubed-ojdk — APPROVED** [review]

File: src/hotspot/share/runtime/continuation.hpp (hunk 2 above)
Comment (AlanBateman) [reply src/hotspot/share/runtime/continuation.hpp:42]: That's right. Additionally, the alternative implementation of virtual threads can be tested by running tests with -XX:-VMContinuations, and we have a few tests that do this.

**Pull request closed without merging**